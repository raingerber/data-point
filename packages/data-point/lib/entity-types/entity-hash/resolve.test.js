/* eslint-env jest */

const resolveHashEntity = require('./resolve').resolve

const FixtureStore = require('../../../test/utils/fixture-store')
const testData = require('../../../test/data.json')

const helpers = require('../../helpers')

let dataPoint
let resolveReducerBound

function transform (entityId, value, options) {
  const reducer = dataPoint.entities.get(entityId)
  const accumulator = helpers.createAccumulator(
    value,
    Object.assign(
      {
        context: reducer
      },
      options
    )
  )
  return resolveHashEntity(accumulator, resolveReducerBound)
}

beforeAll(() => {
  dataPoint = FixtureStore.create()
  resolveReducerBound = helpers.createReducerResolver(dataPoint)
})

describe('Hash entity type checking', () => {
  function resolveInvalid (entity, data) {
    return dataPoint
      .resolve(entity, data)
      .catch(err => err)
      .then(result => {
        expect(result).toBeInstanceOf(Error)
        expect(result).toMatchSnapshot()
      })
  }
  test('should throw error from default outputType reducer when output is not valid', () => {
    return resolveInvalid('hash:asIs', [testData])
  })
  test('should throw error from a custom outputType reducer', () => {
    return resolveInvalid('hash:CustomOutputType', {})
  })
  test('should execute the default outputType reducer before a custom outputType reducer ', () => {
    return resolveInvalid('hash:CustomOutputType', [])
  })
})

describe('entity.hash.value', () => {
  test('should resolve contxt Transform', () => {
    return transform('hash:a.1', testData).then(acc => {
      expect(acc.value).toEqual({
        h1: 1,
        h2: 2,
        h3: 3
      })
    })
  })
})

describe('entity.hash.mapKeys', () => {
  test('should map hash to new keys', () => {
    return transform('hash:b.1', testData).then(acc => {
      expect(acc.value).toEqual({ h: 2 })
    })
  })
  test('returns empty object if mapKeys is empty', () => {
    return transform('hash:b.2', testData).then(acc => {
      expect(acc.value).toEqual({})
    })
  })
})

describe('entity.hash.addKeys', () => {
  test('should add new keys to hash', () => {
    return transform('hash:c.1', testData).then(acc => {
      expect(acc.value).toEqual({
        h1: 1,
        h2: 2,
        h3: 3,
        h4: 4
      })
    })
  })
  test('it should do nothing if addKeys is empty', () => {
    return transform('hash:c.2', testData).then(acc => {
      expect(acc.value).toEqual({ g1: 1 })
    })
  })
})

describe('entity.hash.omitKeys', () => {
  test('should omit keys from hash', () => {
    return transform('hash:d.1', testData).then(acc => {
      expect(acc.value).toEqual({
        h3: 3
      })
    })
  })
  test('it should do nothing if omitKeys is empty', () => {
    return transform('hash:d.2', testData).then(acc => {
      expect(acc.value).toEqual({ g1: 1 })
    })
  })
})

describe('entity.hash.pickKeys', () => {
  test('should only pick keys from hash', () => {
    return transform('hash:e.1', testData).then(acc => {
      expect(acc.value).toEqual({
        h1: 1,
        h2: 2
      })
    })
  })
  test('returns empty object if pickKeys is empty', () => {
    return transform('hash:e.2', testData).then(acc => {
      expect(acc.value).toEqual({})
    })
  })
})

describe('entity.hash.addValues', () => {
  test('should add values to hash', () => {
    return transform('hash:f.1', testData).then(acc => {
      expect(acc.value).toEqual({
        h0: 0,
        h1: 1,
        h2: 2,
        h3: 3
      })
    })
  })
  test('it should do nothing if addValues is empty', () => {
    return transform('hash:f.2', testData).then(acc => {
      expect(acc.value).toEqual({ g1: 1 })
    })
  })
})

describe('entity.hash.compose', () => {
  test('should resolved composed modifiers', () => {
    return transform('hash:h.1', testData).then(acc => {
      expect(acc.value).toEqual({
        e3: 'eThree'
      })
    })
  })
})
