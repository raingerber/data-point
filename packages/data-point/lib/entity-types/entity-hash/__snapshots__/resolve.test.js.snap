// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hash entity type checking should execute the default outputType reducer before a custom outputType reducer  1`] = `
[Error: Entity type check failed!
Expected type: object
Actual type: array
Input value: []]
`;

exports[`Hash entity type checking should throw error from a custom outputType reducer 1`] = `[Error: custom type check failed]`;

exports[`Hash entity type checking should throw error from default outputType reducer when output is not valid 1`] = `
[Error: Entity type check failed!
Expected type: object
Actual type: array
Input value: [ { foo: 1, a: { b: [Object...]
`;
